servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,451 @@
+    mEvictionLock.readLock().lock();
+    try {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      try {
+        BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+        BlockStoreLocation oldLocation = blockMeta.getBlockLocation();
+        moveBlockNoLock(blockId, newLocation);
+        blockMeta = mMetaManager.getBlockMeta(blockId);
+        BlockStoreLocation actualNewLocation = blockMeta.getBlockLocation();
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onMoveBlockByClient(userId, blockId, oldLocation, actualNewLocation);
+        }
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void removeBlock(long userId, long blockId) throws IOException {
+    mEvictionLock.readLock().lock();
+    try {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      try {
+        removeBlockNoLock(userId, blockId);
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onRemoveBlockByClient(userId, blockId);
+        }
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void accessBlock(long userId, long blockId) {
+    for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+      listener.onAccessBlock(userId, blockId);
+    }
+  }
+
+  @Override
+  public void freeSpace(long userId, long availableBytes, BlockStoreLocation location)
+      throws IOException {
+    mEvictionLock.writeLock().lock();
+    try {
+      freeSpaceInternal(userId, availableBytes, location);
+    } finally {
+      mEvictionLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public void cleanupUser(long userId) throws IOException {
+    mEvictionLock.readLock().lock();
+    List<TempBlockMeta> tempBlocksToRemove = mMetaManager.cleanupUser(userId);
+    mLockManager.cleanupUser(userId);
+    mEvictionLock.readLock().unlock();
+
+    // TODO: fix the block removing below,
+    Set<String> dirs = new HashSet<String>(tempBlocksToRemove.size());
+    for (TempBlockMeta tempBlockMeta : tempBlocksToRemove) {
+      String fileName = tempBlockMeta.getPath();
+      try {
+        String dirName = CommonUtils.getParent(fileName);
+        dirs.add(dirName);
+      } catch (InvalidPathException e) {
+        LOG.error("Error in cleanup userId {}: cannot parse parent dir of {}", userId, fileName);
+      }
+      if (!new File(fileName).delete()) {
+        LOG.error("Error in cleanup userId {}: cannot delete file {}", userId, fileName);
+      }
+    }
+    // Cleanup the user folder
+    for (String dirName : dirs) {
```

Comment:
The user folder will not be deleted if there are no temp blocks. We should just iterate through the possible user folders (ie. storage dir/user id), check if it exists, and if it does delete it.